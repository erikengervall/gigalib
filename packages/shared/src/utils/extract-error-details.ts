type ErrorDetails = {
  errorMessage: string;
  errorName: string;
  errorStack?: string;
};

const GENERIC_ERROR_MESSAGE = 'An error occurred';

export function extractErrorDetails(error: unknown): ErrorDetails {
  if (error instanceof Error) {
    return {
      errorMessage: error.message,
      errorName: error.name,
      errorStack: error.stack,
    };
  }

  if (!error) {
    return {
      errorMessage: 'Unknown error (with empty content)',
      errorName: GENERIC_ERROR_MESSAGE,
    };
  }

  if (typeof error !== 'string' && typeof error !== 'object') {
    return {
      errorMessage: `${error ?? 'Unknown error (with unknown type)'}`,
      errorName: GENERIC_ERROR_MESSAGE,
    };
  }

  return {
    errorMessage: JSON.stringify(error),
    errorName: GENERIC_ERROR_MESSAGE,
  };
}
