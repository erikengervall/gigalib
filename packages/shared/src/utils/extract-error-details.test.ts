import { extractErrorDetails } from './extract-error-details';

describe('extractErrorDetails', () => {
  describe('known errors', () => {
    it('should return the error name and message for Error', () => {
      const error = new Error('This is an error');

      const result = extractErrorDetails(error);

      expect(result).toEqual({
        errorName: 'Error',
        errorMessage: 'This is an error',
        errorStack: error.stack,
      });
    });

    it('should return the error name and message for custom Error class', () => {
      class CustomError extends Error {}
      const customError = new CustomError('This is an error');

      const result = extractErrorDetails(customError);

      expect(result).toEqual({
        errorName: 'Error',
        errorMessage: 'This is an error',
        errorStack: customError.stack,
      });
    });

    it('should handle custom error with "errorMessage" property', () => {
      class ErrorWithErrorMessage extends Error {
        errorMessage: string;
        constructor(message: string) {
          super();
          this.errorMessage = message;
        }
      }

      const errorMessage = 'potato with message';
      const error = new ErrorWithErrorMessage(errorMessage);
      const result = extractErrorDetails(error);

      expect(result).toEqual({
        errorName: 'Error',
        errorMessage: errorMessage,
        errorStack: error.stack,
      });
    });

    it('should handle custom error with "internalErrorMessage" property', () => {
      class ErrorWithInternalErrorMessage extends Error {
        internalErrorMessage: string;
        constructor(message: string) {
          super();
          this.internalErrorMessage = message;
        }
      }

      const errorMessage = 'potato with internal message';
      const error = new ErrorWithInternalErrorMessage(errorMessage);
      const result = extractErrorDetails(error);

      expect(result).toEqual({
        errorName: 'Error',
        errorMessage: errorMessage,
        errorStack: error.stack,
      });
    });

    it('should handle custom error with "externalErrorMessage" property', () => {
      class ErrorWithExternalErrorMessage extends Error {
        externalErrorMessage: string;
        constructor(message: string) {
          super();
          this.externalErrorMessage = message;
        }
      }

      const errorMessage = 'potato with external message';
      const error = new ErrorWithExternalErrorMessage(errorMessage);
      const result = extractErrorDetails(error);

      expect(result).toEqual({
        errorName: 'Error',
        errorMessage: errorMessage,
        errorStack: error.stack,
      });
    });
  });

  describe('unknown errors', () => {
    describe('with empty content', () => {
      it('should return the error name and message when the error is an empty string', () => {
        const error = '';

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is null', () => {
        const error = null;

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is undefined', () => {
        const error = undefined;

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });
    });

    describe('with content', () => {
      it('should return the error name and message when the error is string', () => {
        const error = 'This is an error';

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is a number', () => {
        const error = 123;

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is a boolean', () => {
        const error = true;

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is an object', () => {
        const error = {
          foo: 'bar',
        };

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is an array', () => {
        const error = ['foo', 'bar'];

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });

      it('should return the error name and message when the error is a function', () => {
        const error = () => void 0;

        const result = extractErrorDetails(error);

        expect(result).toMatchSnapshot();
      });
    });
  });
});
