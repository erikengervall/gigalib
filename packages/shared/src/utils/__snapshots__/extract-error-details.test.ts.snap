// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is a boolean 1`] = `
{
  "errorMessage": "true",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is a function 1`] = `
{
  "errorMessage": "() => void 0",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is a number 1`] = `
{
  "errorMessage": "123",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is an array 1`] = `
{
  "errorMessage": "["foo","bar"]",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is an object 1`] = `
{
  "errorMessage": "{"foo":"bar"}",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with content should return the error name and message when the error is string 1`] = `
{
  "errorMessage": ""This is an error"",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with empty content should return the error name and message when the error is an empty string 1`] = `
{
  "errorMessage": "Unknown error (with empty content)",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with empty content should return the error name and message when the error is null 1`] = `
{
  "errorMessage": "Unknown error (with empty content)",
  "errorName": "An error occurred",
}
`;

exports[`extractErrorDetails unknown errors with empty content should return the error name and message when the error is undefined 1`] = `
{
  "errorMessage": "Unknown error (with empty content)",
  "errorName": "An error occurred",
}
`;
